tf2 Tutorial: Adding fixed and dynamic frames (Python) (#1737)
ros2/ros2_documentation: source/Tutorials/Tf2/Adding-A-Frame-Py.rst

1.2 Build and run
~~~~~~~~~~~~~~~~~

Edit the ``turtle_tf2_demo.launch.py`` launch file. Simply add the following line:

.. code-block:: python

   from launch import LaunchDescription
   from launch_ros.actions import Node

   def generate_launch_description():
      return LaunchDescription([
         ...,
         Node(
            package='learning_tf2_py',
            executable='fixed_frame_tf2_broadcaster',
            name='fixed_broadcaster',
         ),
      ])

This will add our fixed ``carrot1`` frame to the turtlesim world.
Finally, start the turtle broadcaster demo:

.. code-block:: console

   ros2 launch learning_tf2_py turtle_tf2_demo.launch.py

You should also notice that the new ``carrot1`` frame appeared in the transformation tree.

.. image:: turtlesim_frames_carrot.png

1.3 Checking the results
~~~~~~~~~~~~~~~~~~~~~~~~

So, if you drive the first turtle around, you notice that the behavior didn't change from the previous tutorial, even though we added a new frame.
That's because adding an extra frame does not affect the other frames, and our listener is still using the previously defined frames.
So, let's change the behavior of the listener.

> Open the ``turtle_tf2_listener.py`` file, and simply replace ``turtle1`` with ``carrot1`` on line 62:

Reviewer: can you highlight this an important think to do ?

Author: I updated the explanation part. Is it more clear now?

.. _AddingAFramePy:

Adding a frame (Python)

> =========================================

Reviewer 2: Nit: make this the same length as the title.

**Goal:** Learn how to to add an extra frame to tf2.

**Tutorial level:** Intermediate

**Time:** 10 minutes

.. contents:: Contents
   :depth: 2
   :local:

Background
----------

In previous tutorials, we recreated the turtle demo by writing a :ref:`tf2 broadcaster <WritingATf2BroadcasterPy>` and a :ref:`tf2 listener <WritingATf2ListenerPy>`.
This tutorial will teach you how to add an extra frame to the transformation tree.
In fact, adding a frame in tf2 is very similar to creating the tf2 broadcaster, but this example will show you some additional features of tf2.

> 1 Why adding frames
> ^^^^^^^^^^^^^^^^^^^

Reviewer 2: ```suggestion
1 Why add frames?
^^^^^^^^^^^^^^^^^
```

For many tasks related to transformations, it is easier to think inside a local frame.
For example, it is easiest to reason about laser scan measurements in a frame at the center of the laser scanner.
tf2 allows you to define a local frame for each sensor, link, or joint in your system.

> Finally, tf2 will take care of all the hidden intermediate frame transformations that are introduced.

Reviewer 2: ```suggestion
When transforming from one frame to another, tf2 will take care of all the hidden intermediate frame transformations that are introduced.
```

2 Where to add frames
^^^^^^^^^^^^^^^^^^^^^

> tf2 builds up a tree structure of frames, and thus it does not allow a closed loop in the frame structure.

Reviewer 2: ```suggestion
tf2 builds up a tree structure of frames, and thus does not allow a closed loop in the frame structure.
```

This means that a frame only has one single parent, but it can have multiple children.
Currently, our tf2 tree contains three frames: ``world``, ``turtle1`` and ``turtle2``.
The two turtle frames are children of the ``world`` frame.
If we want to add a new frame to tf2, one of the three existing frames needs to be the parent frame, and the new one will become its child frame.

.. image:: turtlesim_frames.png

Tasks
-----

1 Write the fixed frame broadcaster
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In our turtle example, we'll add a new frame ``carrot1``, which will be the child of the ``turtle1``. This frame will serve as the goal for the second turtle.

> Let's first create the source files. Go to the ``learning_tf2_py`` package we created in the previous tutorials.

Reviewer 2: Nit: one sentence per line.

         t.transform.translation.y = 2.0
         t.transform.translation.z = 0.0
         t.transform.rotation.x = 0.0
         t.transform.rotation.y = 0.0
         t.transform.rotation.z = 0.0
         t.transform.rotation.w = 1.0

         self.br.sendTransform(t)

   def main():
      rclpy.init()
      node = FixedFrameBroadcaster()
      try:
         rclpy.spin(node)
      except KeyboardInterrupt:
         pass

      rclpy.shutdown()

Don't forget to add the node to the ``setup.py``.
The code is very similar to the tf2 broadcaster tutorial example and the only difference is that the transform here does not change over time.

1.1 Examine the code
~~~~~~~~~~~~~~~~~~~~

Let's take a look at the key lines in this piece of code.
Here we create a new transform, from the parent ``turtle1`` to the new child ``carrot1``.
The ``carrot1`` frame is 2 meters offset in y axis in terms of the ``turtle1`` frame.

.. code-block:: python

   t = TransformStamped()
   t.header.stamp = self.get_clock().now().to_msg()
   t.header.frame_id = 'turtle1'
   t.child_frame_id = 'carrot1'
   t.transform.translation.x = 0.0
   t.transform.translation.y = 2.0
   t.transform.translation.z = 0.0

> 1.2 Build and run

Reviewer 2: It feels to me like this section should be "Create the launch file", or something like that.  Then Section 1.3 would be "Building the workspace", and section 1.4 would be "Run the launch file and check the results".

Author: Fixed in https://github.com/ros2/ros2_documentation/pull/1737/commits/fe559ba438dfc7475b1c4132b869a946e02af1de.
Does it look better this way?

~~~~~~~~~~~~~~~~~

> Now let's create a launch file for this example. With your text editor, create a new

Reviewer 2: Nit: one sentence per line.

file called ``turtle_tf2_fixed_frame_demo.launch.py``, and add the following lines:

.. code-block:: python

   import os

   from ament_index_python.packages import get_package_share_directory

   from launch import LaunchDescription
   from launch.actions import IncludeLaunchDescription
   from launch.launch_description_sources import PythonLaunchDescriptionSource

   from launch_ros.actions import Node

   def generate_launch_description():
      demo_nodes = IncludeLaunchDescription(
         PythonLaunchDescriptionSource([os.path.join(
               get_package_share_directory('learning_tf2_py'), 'launch'),
               '/turtle_tf2_demo.launch.py']),
         )

      return LaunchDescription([
         demo_nodes,
         Node(
               package='learning_tf2_py',
               executable='fixed_frame_tf2_broadcaster',
               name='fixed_broadcaster',
         ),
      ])

> This launch file will first import the required packages and create a ``demo_nodes`` variable that will store
> nodes that we created in the previous tutorial's launch file.

Reviewer 2: ```suggestion
This launch file first imports the required packages, then creates a ``demo_nodes`` variable that will store
nodes that we created in the previous tutorial's launch file.
```

The last part of the code will add our fixed ``carrot1`` frame to the turtlesim world using our ``fixed_frame_tf2_broadcaster`` node.

.. code-block:: python

   Node(
      package='learning_tf2_py',
      executable='fixed_frame_tf2_broadcaster',
      name='fixed_broadcaster',
   ),

Now start the turtle broadcaster demo:

.. code-block:: console

   ros2 launch learning_tf2_py turtle_tf2_fixed_frame_demo.launch.py

You should notice that the new ``carrot1`` frame appeared in the transformation tree.

.. image:: turtlesim_frames_carrot.png

1.3 Checking the results
~~~~~~~~~~~~~~~~~~~~~~~~

> So, if you drive the first turtle around, you notice that the behavior didn't change from the previous tutorial, even though we added a new frame.

Reviewer 2: ```suggestion
If you drive the first turtle around, you should notice that the behavior didn't change from the previous tutorial, even though we added a new frame.
```

         x = seconds * math.pi

         t = TransformStamped()
         t.header.stamp = self.get_clock().now().to_msg()
         t.header.frame_id = 'turtle1'
         t.child_frame_id = 'carrot1'
         t.transform.translation.x = 10 * math.sin(x)
         t.transform.translation.y = 10 * math.cos(x)
         t.transform.translation.z = 0.0
         t.transform.rotation.x = 0.0
         t.transform.rotation.y = 0.0
         t.transform.rotation.z = 0.0
         t.transform.rotation.w = 1.0

         self.br.sendTransform(t)

   def main():
      rclpy.init()
      node = DynamicFrameBroadcaster()
      try:
         rclpy.spin(node)
      except KeyboardInterrupt:
         pass

      rclpy.shutdown()

2.1 Examine the code
~~~~~~~~~~~~~~~~~~~~

Instead of a fixed definition of our x and y offsets, we are using the ``sin()`` and ``cos()`` functions on the current time so that the offset of ``carrot1`` is constantly changing.

.. code-block:: python

   seconds, _ = self.get_clock().now().seconds_nanoseconds()
   x = seconds * math.pi
   ...
   t.transform.translation.x = 10 * math.sin(x)
   t.transform.translation.y = 10 * math.cos(x)

> 2.2 Build and run

Reviewer 2: Same note here about splitting this up into "Create the launch file", build, run, and examine the results.

Author: Fixed in https://github.com/ros2/ros2_documentation/pull/1737/commits/fe559ba438dfc7475b1c4132b869a946e02af1de.

2.1 Examine the code
~~~~~~~~~~~~~~~~~~~~

Instead of a fixed definition of our x and y offsets, we are using the ``sin()`` and ``cos()`` functions on the current time so that the offset of ``carrot1`` is constantly changing.

.. code-block:: python

   seconds, _ = self.get_clock().now().seconds_nanoseconds()
   x = seconds * math.pi
   ...
   t.transform.translation.x = 10 * math.sin(x)
   t.transform.translation.y = 10 * math.cos(x)

2.2 Write the launch file
~~~~~~~~~~~~~~~~~~~~~~~~~

To test this code, create a new launch file ``turtle_tf2_dynamic_frame_demo.launch.py`` and paste the following code:

.. code-block:: python

   import os

   from ament_index_python.packages import get_package_share_directory

   from launch import LaunchDescription
   from launch.actions import IncludeLaunchDescription
   from launch.launch_description_sources import PythonLaunchDescriptionSource

   from launch_ros.actions import Node

   def generate_launch_description():
      demo_nodes = IncludeLaunchDescription(
         PythonLaunchDescriptionSource([os.path.join(
               get_package_share_directory('learning_tf2_py'), 'launch'),
               '/turtle_tf2_demo.launch.py']),

> 

Reviewer: ```suggestion
```
